File: src/NodeParser/InferTypeNodeParser.ts
import ts from "typescript";
import type { Context, NodeParser } from "../NodeParser.js";
import type { SubNodeParser } from "../SubNodeParser.js";
import type { BaseType } from "../Type/BaseType.js";
import { InferType } from "../Type/InferType.js";

export class InferTypeNodeParser implements SubNodeParser {
    public constructor(
        protected typeChecker: ts.TypeChecker,
        protected childNodeParser: NodeParser,
    ) {}

    public supportsNode(node: ts.InferTypeNode): boolean {
        return node.kind === ts.SyntaxKind.InferType;
    }

    public createType(node: ts.InferTypeNode, _context: Context): BaseType {
        return new InferType(node.typeParameter.name.escapedText.toString());
    }
}

File: internal/parser/infer_type_node_parser.go
package parser

import (
	"github.com/microsoft/typescript-go/shim/ast"
	"github.com/microsoft/typescript-go/shim/checker"

	"github.com/vega/ts-json-schema-generator/internal/types"
)

// InferTypeNodeParser handles `infer T` placeholders which are resolved later
// by the conditional type parser (src/NodeParser/InferTypeNodeParser.ts).
type InferTypeNodeParser struct {
	typeChecker     *checker.Checker
	childNodeParser NodeParser
}

func NewInferTypeNodeParser(typeChecker *checker.Checker, childNodeParser NodeParser) *InferTypeNodeParser {
	return &InferTypeNodeParser{typeChecker: typeChecker, childNodeParser: childNodeParser}
}

func (p *InferTypeNodeParser) SupportsNode(node *ast.Node) bool {
	return node.Kind == ast.KindInferType
}

func (p *InferTypeNodeParser) CreateType(node *ast.Node, _ *Context, _ *types.ReferenceType) types.Type {
	return types.NewInferType(node.AsInferTypeNode().TypeParameter.Name().Text())
}
